proposals/stdlib/success-or-failure.md

```diff
@@ -0,0 +1,687 @@
+    runCatching { await() }
+```
+
+### Functional bulk manipulation of failures
+
+Kotlin encourages writing code in a functional style. It works well as long as business-specific failures are
+represented with nullable types or sealed class hierarchies, while other kinds of failures 
+(that are represented by exceptions) do not require any special local handling. However, when interfacing with
+Java-style APIs that rely heavily on exceptions or otherwise having a need to somehow process exceptions locally
+(as opposed to propagating them up the call stack), we see a clear lack of primitives in the Kotlin standard library.
+
+Consider writing a function `readFiles` that receives a list of files, reads all of them, and returns a 
+list of results. We are given the following function to read single file contents:
+
+```kotlin
+fun readFileData(file: File): Data
+```
+
+This reading function throws exception if file is not found or parsing of a file had somehow failed. Normally that would
+be fine and the first failure of this kind would terminate the whole program with a stacktrace and explanatory message. 
+However, for `readFiles` we'd explicitly like to be able to continue after the failure to collect and report all failures.
+Moreover, we'd like to be able to have a functional implementation of `readFiles` like this:
+
+```kotlin
+fun readFilesCatching(files: List<File>): List<SuccessOrFailure<Data>> =
+    files.map { 
+        runCatching { 
+            readFileData(it)
+        }
+    }
+```
+
+> This function is named `readFileCatching` to make it explicit to the caller that all encountered failures
+were _caught_ and encapsulated in `SuccessOrFailure` and it is caller responsibility to process these failures.
+
+Now, consider making some transformation of `readFilesCatching` results that we'd like to express functionally, 
+while preserving accumulated failures:
+
+```kotlin                                                     
+readFilesCatching(files).map { result: SuccessOrFailure<Data> -> // type explicitly written here for clarity
+    result.map { it.doSomething() } // Operates on Success case, while preserving Failure
+}
+```
+
+If `doSomething`, in turn, can potentially fail and we are interested in keeping this failure per each individual
+file, then we can write it using `mapCatching` instead of `map`:
+
+```kotlin
+readFilesCatching(files).map { result: SuccessOrFailure<Data> -> 
+    result.mapCatching { it.doSomething() }
+}
+```
+
+### Functional error handling
+
+In mostly functional code `try { ... } catch(e: Throwable) { ... }` construct looks
+out of style. For example, consider this piece of code that uses [RxKotlin](https://github.com/ReactiveX/RxKotlin) 
+for asynchronous processing.
+It invokes `doSomethingAsync` that returns `Single<Data>` and processes potential error in a functional style:
+
+```kotlin
+doSomethingAsync()
+    .doOnError { showErrorDialog(it) }
```

Comment:
note: this doesn't really handle errors in rx, it's just a side-effect callback. 

In rx errors can be handled by switching to another value/stream or by the final Observer at the end. Otherwise they'll be propagated as exceptions (in RxJava2 they go to [`RxJavaPlugins.onError()`](http://reactivex.io/RxJava/2.x/javadoc/io/reactivex/plugins/RxJavaPlugins.html#onError-java.lang.Throwable-))

Reply:
@artem-zinnatullin Good point. What would an idiomatic Rx piece of code that I can use here?

Reply:
Well, it'll break your comparison down below a bit but since you want to do some final action on the value/error it'll be something like this:

```kotlin
doSomethingAsync()
    .subscribe(
        { processData(it) },
        { showErrorDialog(it) }
    )
```

Please also note that even though RxJava and [Reactive Streams spec](https://github.com/reactive-streams/reactive-streams-jvm) support error (`Throwable`) handling for exceptions thrown in the streams, that's not very scalable way of handling errors in rx world.
Especially on larger projects, regular rx error handling allows bugs to slip to production due to error type information loss. Ie you don't know if you need to handle error on particular rx stream or not unless it's written in the javadoc or you've read its source code.

Modern projects convert expected subset of errors (`Throwable`s) to values of the stream (works especially well with Kotlin's sealed classes).

In a way it's like returning Java's `throws` back to the codebase, but with a better control over types.

Reply:
Thanks. I've updated Rx example and added more explanatory details there. I'm also glad to hear that using sealed classes for business errors is getting tracking. Maybe we could improve something in the language (in the future) to make _that_ pattern easier to use.